### ISSUE TITLE
Scudcloud No Longer Loads / Logs In

### ISSUE THREAD

**Issue by stratigos** [raelgc/scudcloud#551]
## ScudCloud Version

Paste the output for `scudcloud --version` below:
```
ScudCloud 1.50

Python 3.4.3

Qt 5.2.1

PyQt 5.2.1

SIP 4.15.5
```

## Distro and Desktop info

  - Distribution: Ubuntu 14.04
  - Version/Release: 14.04 LTS
  - Desktop environment: Unity

## Expected behavior

I click the scudcloud icon, slack loads and logs me into my team.

## Actual behavior

I click scudcloud, sign into my team, slack hangs at loading screen forever.

## Steps to reproduce
 1. Click scudcloud icon
 2. log into team, press enter key
 3. screen changes to loading screen, animation stops, "Trouble? Run the self help test" link at the bottom does nothing. 

----

Here is debug output:

```
scudcloud --debug DEBUG

URL Changed: https://slack.com/signin

URL Changed: https://MYTEAM.slack.com/messages
```

...thats it. 

I have completely uninstalled and reinstalled the application to no avail. I wiped my `~./config/scudcloud` dir as well.

**Comment by matt-lindsay** [comment]
Hi @stratigos after an aptitude update / upgrade yesterday I'm getting the same: Scudcloud just hangs. I'm using the following...

ScudCloud 1.50
Python 3.4.3
Qt 5.2.1
PyQt 5.2.1
SIP 4.15.5

My OS is Mint 17.3

**Comment by mrlphoenix** [comment]
Same thing on Ubuntu 15.10
ScudCloud 1.37
Python 3.4.3+
Qt 4.8.6
PyQt 4.11.4
SIP 4.16.9

__main__.py is already fixed.

**Comment by raelgc** [comment]
@SpaceWind You're using a no more supported version of Ubuntu. I cannot publish new versions of Scudcloud for it.

**Comment by ben-cohen** [comment]
I have the same problem on Linux Mint 17 Qiana.

> ScudCloud 1.50
> Python 3.4.3
> Qt 5.2.1
> PyQt 5.2.1
> SIP 4.15.5

**Comment by mocrel** [comment]
I have the same issue. Xubuntu 14.04 LTS. 

$ scudcloud --version
ScudCloud 1.50
Python 3.4.3
Qt 5.2.1
PyQt 5.2.1
SIP 4.15.5

$ uname -a
Linux carbon 3.13.0-112-generic #159-Ubuntu SMP Fri Mar 3 15:27:22 UTC 2017 i686 i686 i686 GNU/Linux

$ python3 -c "import ssl; print(ssl.OPENSSL_VERSION)"
OpenSSL 1.0.1f 6 Jan 2014

**Comment by raelgc** [comment]
The pattern here appears to be Ubuntu 14.04, which uses a older version of Qt5. Any 14.04 experienced user here want to try upgrade Qt5?

**Comment by stratigos** [comment]
I am not the most informed on this issue, I am learning as I go, so please bear with me here. It appears that [the default package for Qt5 in Ubuntu 14.04](http://packages.ubuntu.com/trusty/qt5-default) is set to Qt 5.2.1 (I believe the package is `qt-default`).

----

Brief story:

_TLDR: Upgrading QT doesnt look simple, and Slack has an official client now :-/_

Ive spent some time looking into what it would take to upgrade Qt outside of the default package manager, and I do at the least conclude its not trivial (though may be easier than I suspect). Information on this is a bit scattered and it is easy to find [outdated guides](https://wiki.qt.io/Install_Qt_5_on_Ubuntu#Introduction), though [this tutorial](http://www.bogotobogo.com/Qt/Qt5_Install_Ubuntu_14_64bit.php) looked promising. Unfortunately, it contains an outdated link to download QT. The information on [QT's download page](https://www.qt.io/download/) now contains a lot of language pertaining to purchasing licenses and whatnot, and there is either no information on what I need, or its very hard to find. Heck, Im not even 100% sure what I am looking for. I explored hacking my package manager [sort of like in this stackoverflow post](http://stackoverflow.com/questions/12600661/how-to-upgrade-qt-installed-in-linux-from-one-version-to-higher), but had difficulty locating the exact source I need to update and from where is the best place to get it. Running `dpkg --list | grep -i qt` yielded a ton of results, and "lots of things" depending on QT 5.1.2 - no idea what the consequences of updating one or many dependencies is. 

Meanwhile, Slack now has an official Linux client. Scudcloud is better in my opinion, as the official Slack Linux client lacks Unity launcher notifications, and has extremely subtle Unity panel notifications, resulting in me not noticing chats I receive when not looking at my screen(s). I also like how scudcloud handles multiple teams more than the Slack official. I'd rather use Scudcloud, but not with the cognitive burden that the above presents to me. I have this familiar feeling in the back of my mind that "figuring out the QT situation for 14.04" is going to lead to temporary loss of my favorite business tools.

So Im going to just go with the official Slack client on this machine for now. Maybe someone else has the bandwidth or knowledge on how to fix this situation for us Ubuntu 14.04 Scudcloud users, and I can come back later (note: I also have an Ubuntu 16.04 machine, and will be using Scudcloud on that). Believe me, I'd rather use ScudCloud :). 

----

_TLDR: I dont feel like upgrading my system either._

I should probably upgrade Ubuntu, but Ive actually been holding out for 18.04 next year. Ubuntu 14.04 is still supported LTS, and I have _zero_ other conflicts regarding my version of Ubuntu and any other software I use. I have been an Ubuntu user for a long time, and perhaps things are better now, but my memories of upgrading the system are very painful - I never upgrade anymore, I always fresh install to a new hard disk and copy my home dir. History taught me this is the fastest and easiest way. Thus I am in no hurry to upgrade from Ubuntu 14.04, and I have 16.04 on my secondary work machine, and will continue to use Scudcloud there.

**Comment by raelgc** [comment]
Ok. Fair enough.

Let's try figure what is prevening Scudcloud get loaded in 14.04.

Please, start scudcloud with `scudcloud --debug=True &`. Then right click in the main (messages) pane, and select `Inspect`. Maximize the window and select `Console` tab.

We should see some Slack errors (expected) and some errors preventing Scudcloud to get fully loaded.

Please, paste the errors here.

**Comment by mocrel** [comment]
Here are the errors that I get ...

![sc](https://cloud.githubusercontent.com/assets/26916561/24867247/e88ee79c-1e0c-11e7-8481-bc3e8ca9528e.png)

**Comment by raelgc** [comment]
Hm, this is easier than I though. I'll fix it in next couple of hours.

Thanks @mocrel for this awesome feedback!

**Comment by raelgc** [comment]
Fixed and published for Ubuntu 14.04.

@SpaceWind Can you try the debug steps and see if the `MutationObserver` error is present for you too?

**Comment by mocrel** [comment]
Yes, your fix worked, everything is fine again! Thanks a million, @raelgc !!!

**Comment by raelgc** [comment]
Welcome and thank you for the help!

**Comment by ben-cohen** [comment]
@raelgc I see the same error as @mocrel in the Inspector: "e: ReferenceError: Can't find variable: MutationObserver".

Your commit b8f26828d fixes it for me most of the time - thank you -  but I do still get the hang and the same error sometimes.  (When it succeeds that error doesn't appear.)

**Comment by ben-cohen** [comment]
I tried putting this at the start of scudcloud.js:

```javascript
console.log("MutationObserver " + typeof MutationObserver);
console.log("WebKitMutationObserver " + typeof WebKitMutationObserver);
var MutationObserver = MutationObserver || WebKitMutationObserver;
console.log("MutationObserver result " + typeof MutationObserver);
```
Then on success I saw this output in the Inspector Console log:
```
...
2017/4/10 19:41:38.610  Setting calling_rtm_start to true rollup-core_required_ts.js:1
2017/4/10 19:41:38.610  Starting incremental boot rollup-core_required_ts.js:1
MutationObserver undefined
WebKitMutationObserver object
MutationObserver result object
TypeError: 'undefined' is not an object (evaluating 'model_ob_then.id') rollup-client.js:8
ScudCloud.didStartLoading, args: 60000
ScudCloud.inputFieldCreated, args: 
...
```

On failure I saw the same output from the log, but the "ReferenceError: Can't find variable: MutationObserver" message appears before that.  Therefore the problem is that the MutationObserver is being referenced too soon.

I tried adding the MutationObserver assignment to the end of `Wrapper.__init__()`.  (I don't know whether that's a good place to put it.)
```python
...
class Wrapper(QWebView):
    ...
    def __init__(self, window):
        ...
        self.addActions()
        self.page().currentFrame().evaluateJavaScript("var MutationObserver = MutationObserver || WebKitMutationObserver;")
self.page().currentFrame().evaluateJavaScript("var MutationObserver = MutationObserver || WebKitMutationObserver;")

    def permissionRequested(self, frame, feature):
        ...
    ...
...
```

Now the line printed in the log has changed to "MutationObserver object" and it has connected successfully every time since then.  (With that you don't need the change to scudcloud.js.)

**Comment by raelgc** [comment]
Nice one, @ben-cohen!

Can you, please, create a pull request with this fix?

**Comment by raelgc** [comment]
@ben-cohen Maybe a better place is inside the `_loadStarted(self):` method. Can you test, please?

**Comment by ben-cohen** [comment]
@raelgc I have just tested it in `Wrapper._loadStarted(self)` and it works.  I see you have already committed that change in commit de5eae87 so I haven't created a PR.  Thank you for the fix. :)

**Comment by raelgc** [comment]
@ben-cohen Welcome and thank you for investigate!

**Comment by stratigos** [comment]
Today I ran `sudo apt-get upgrade` on my Ubuntu 14.04 machine, and noticed `scudcloud` was in the list. After the package upgrade, I am able to log in to my team again! Thanks for all your hard work, @raelgc.
